examples/vision/semantic_image_clustering.py

```diff
@@ -0,0 +1,577 @@
+image = x_data[image_idx]
+image_class = class_labels[y_data[image_idx][0]]
+plt.figure(figsize=(3, 3))
+plt.imshow(x_data[image_idx].astype("uint8"))
+plt.title(image_class)
+_ = plt.axis("off")
+
+"""
+Display a sample of augmented versions of the image
+"""
+
+plt.figure(figsize=(10, 10))
+for i in range(9):
+    augmented_images = data_augmentation(np.array([image]))
+    ax = plt.subplot(3, 3, i + 1)
+    plt.imshow(augmented_images[0].numpy().astype("uint8"))
+    plt.axis("off")
+
+"""
+## Self-supervised representation learning
+"""
+
+"""
+### Implement the vision encoder
+"""
+
+
+def create_encoder(representation_dim):
+    encoder = keras.Sequential(
+        [
+            keras.applications.ResNet50V2(
+                include_top=False, weights=None, pooling="avg"
+            ),
+            layers.Dense(representation_dim),
+        ]
+    )
+    return encoder
+
+
+"""
+### Implement the unsupervised contrastive loss
+"""
+
+
+class ContrastiveLoss(keras.losses.Loss):
+    def __init__(self, temperature=1.0, l2_normalize=False):
+        super(ContrastiveLoss, self).__init__()
+        self.temperature = temperature
+        self.l2_normalize = l2_normalize
+
+    def __call__(self, labels, feature_vectors, sample_weight=None):
+        # The shape of feature_vectors is [num_augmentations * batch_size, projection_units].
+        # Convert labels to an identiy matrix of shape [batch_size, batch_size].
+        labels = tf.eye(tf.shape(labels)[0])
+        if self.l2_normalize:
+            feature_vectors = tf.math.l2_normalize(feature_vectors, -1)
+        # The logits shape is [num_augmentations * batch_size, num_augmentations * batch_size].
+        logits = (
+            tf.linalg.matmul(feature_vectors, feature_vectors, transpose_b=True)
+            / self.temperature
+        )
+        # Apply log-max trick for numerical stability.
+        logits_max = tf.math.reduce_max(logits, axis=1)
+        logits = logits - logits_max
+        # Compute num_augmentations.
+        num_augmentations = tf.shape(feature_vectors)[0] // tf.shape(labels)[0]
+        # The shape of targets is [num_augmentations * batch_size, num_augmentations * batch_size].
+        targets = tf.tile(labels, [num_augmentations, num_augmentations])
+        # Compute cross entropy loss
+        loss = keras.losses.categorical_crossentropy(
+            y_true=targets, y_pred=logits, from_logits=True
+        )
+        return loss
+
+
+"""
+### Implement representation learner
+"""
+
+
+def create_representation_learner(encoder, num_augumentations, projection_units):
+
+    inputs = keras.Input(shape=input_shape)
+    # Preprocess the input images.
+    preprocessed = data_preprocessing(inputs)
+    # Create augmented versions of the images.
+    augmented = []
+    for _ in range(num_augumentations):
+        augmented.append(data_augmentation(preprocessed))
+    augmented = layers.Concatenate(axis=0)(augmented)
```

Comment:
Technically this model breaks a fundamental assumption of the API, which is the batch size should be constant throughout the model. Overall it works fine though. A way to avoid this could be to keep the different model parts separate and only combine them in the `train_step` of a `Model` subclass.